src/rules/no-extraneous-dependencies.js

```diff
@@ -1,86 +1,89 @@
+    peerDependencies,
   }
 }
 
-function getDependencies(context, packageDir) {
-  let paths = []
+function assignDeps(fromDeps, toDeps) {
+  Object.assign(fromDeps.dependencies, toDeps.dependencies)
+  Object.assign(fromDeps.devDependencies, toDeps.devDependencies)
+  Object.assign(fromDeps.peerDependencies, toDeps.peerDependencies)
+  Object.assign(fromDeps.optionalDependencies, toDeps.optionalDependencies)
+}
+
+function getDependencies(context, packageDir, filename) {
+  const files = []
+
   try {
-    const packageContent = {
-      dependencies: {},
-      devDependencies: {},
-      optionalDependencies: {},
-      peerDependencies: {},
-    }
+    const closest = readPkgUp.sync({cwd: filename, normalize: false})
+    files.push(closest.path)
 
-    if (!isEmpty(packageDir)) {
-      if (!isArray(packageDir)) {
-        paths = [path.resolve(packageDir)]
-      } else {
-        paths = packageDir.map(dir => path.resolve(dir))
-      }
-    }
+    const deps = (packageDir ? [].concat(packageDir) : [])
+      .reduce((allDeps, dir) => {
+        const pkgFile = path.resolve(dir, 'package.json')
 
-    if (!isEmpty(paths)) {
-      // use rule config to find package.json
-      paths.forEach(dir => {
-        Object.assign(packageContent, extractDepFields(
-          JSON.parse(fs.readFileSync(path.join(dir, 'package.json'), 'utf8'))
-        ))
-      })
-    } else {
-      // use closest package.json
-      Object.assign(
-        packageContent,
-        extractDepFields(
-          readPkgUp.sync({cwd: context.getFilename(), normalize: false}).pkg
-        )
-      )
-    }
+        if (files.indexOf(pkgFile) === -1) {
+          files.push(pkgFile)
+
+          assignDeps(
+            allDeps,
+            extractDepFields(JSON.parse(fs.readFileSync(pkgFile, 'utf8')))
+          )
+        }
+
+        return allDeps
+      }, extractDepFields(closest.pkg))
 
-    if (![
-      packageContent.dependencies,
-      packageContent.devDependencies,
-      packageContent.optionalDependencies,
-      packageContent.peerDependencies,
+    if ([
+      deps.dependencies,
+      deps.devDependencies,
+      deps.optionalDependencies,
+      deps.peerDependencies,
     ].some(hasKeys)) {
-      return null
+      return deps
     }
-
-    return packageContent
   } catch (e) {
-    if (!isEmpty(paths) && e.code === 'ENOENT') {
+    const relFiles = files.map((file) => path.relative(CWD, file))
+
+    if (e.code === 'ENOENT') {
       context.report({
-        message: 'The package.json file could not be found.',
+        message: `Could not find: ${relFiles.join(', ')}`,
         loc: { line: 0, column: 0 },
       })
-    }
-    if (e.name === 'JSONError' || e instanceof SyntaxError) {
+    } else if (e.name === 'JSONError' || e instanceof SyntaxError) {
       context.report({
-        message: 'The package.json file could not be parsed: ' + e.message,
+        message: `Could not parse ${relFiles.pop()}: ${e.message}`,
```

Comment:
it's a bit shady to pop out of this array as part of the message; can it just be `relFiles[relFiles.length - 1]` instead?